enterprise/server/remote_execution/containers/podman/podman.go

```diff
@@ -43,13 +54,114 @@ const (
 	podmanDefaultNetworkIPRange = "10.88.0.0/16"
 	podmanDefaultNetworkGateway = "10.88.0.1"
 	podmanDefaultNetworkBridge  = "cni-podman0"
+
+	optImageRefCacheSize = 1000
 )
 
+type optImageCache struct {
+	cache interfaces.LRU
+}
+
+func (c *optImageCache) get(key string) (string, error) {
+	if v, ok := c.cache.Get(key); ok {
+		return v.(string), nil
+	}
+	return "", nil
+}
+
+func (c *optImageCache) put(key string, optImage string) {
+	c.cache.Add(key, optImage)
+}
+
+func newOptImageCache() (*optImageCache, error) {
+	l, err := lru.NewLRU(&lru.Config{
+		SizeFn: func(value interface{}) int64 {
+			return 1
+		},
+		MaxSize: optImageRefCacheSize,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &optImageCache{cache: l}, nil
+}
+
 type pullStatus struct {
 	mu     *sync.RWMutex
 	pulled bool
 }
 
+type Provider struct {
+	env                   environment.Env
+	imageCacheAuth        *container.ImageCacheAuthenticator
+	buildRoot             string
+	imageStreamingEnabled bool
+	regClient             regpb.RegistryClient
+	optImageCache         *optImageCache
+}
+
+func NewProvider(env environment.Env, imageCacheAuthenticator *container.ImageCacheAuthenticator, buildRoot string) (*Provider, error) {
+	c, err := newOptImageCache()
+	if err != nil {
+		return nil, err
+	}
+	var regClient regpb.RegistryClient
+	if *imageStreamingRegistryGRPCTarget != "" {
+		conn, err := grpc_client.DialTarget(*imageStreamingRegistryGRPCTarget)
+		if err != nil {
+			return nil, err
+		}
+		regClient = regpb.NewRegistryClient(conn)
+	}
+
+	imageStreamingEnabled := *imageStreamingRegistryHTTPTarget != "" && *imageStreamingRegistryGRPCTarget != ""
+	if imageStreamingEnabled {
+		log.Infof("Starting stargz store")
+		cmd := exec.CommandContext(env.GetServerContext(), "stargz-store", "/var/lib/stargz-store/store")
+		logWriter := log.Writer("[stargzstore] ")
+		cmd.Stderr = logWriter
+		cmd.Stdout = logWriter
+		if err := cmd.Start(); err != nil {
+			return nil, status.UnavailableErrorf("could not start stargz store: %s", err)
+		}
+
+		// Configures podman to check stargz store for image data.
+		storageConf := `
+[storage]
+driver = "overlay"
+runroot = "/run/containers/storage"
+graphroot = "/var/lib/containers/storage"
+[storage.options]
+additionallayerstores=["/var/lib/stargz-store/store:ref"]
+`
+		if err := os.WriteFile("/etc/containers/storage.conf", []byte(storageConf), 0644); err != nil {
+			return nil, status.UnavailableErrorf("could not write storage config: %s", err)
+		}
+	}
+
+	return &Provider{
+		env:                   env,
+		imageCacheAuth:        imageCacheAuthenticator,
+		buildRoot:             buildRoot,
+		imageStreamingEnabled: imageStreamingEnabled,
+		regClient:             regClient,
+		optImageCache:         c,
+	}, nil
+}
```

Comment:
Because most of the methods we use are defined on podmanCommandContainer, it seems like we could move most of this code to an initX() error type function on commandContainer and run it behind a sync.Once. That would then let us remove the Provider struct entirely and leave the runner code as-is, which I think would be a little cleaner / simpler?

Reply:
I think having another type responsible for the overall podman configuration is cleaner than hiding the logic behind `sync.Once`.